tests/src/Simple/Threading/Threading.cs

```diff
@@ -758,6 +789,282 @@ public static void MutexMaximumReacquireCountTest()
+        int maxCount = 5;
+        void Job()
+        {
+            if(Interlocked.Increment(ref testJobsCount) != maxCount)
+            {
+                Task.Factory.StartNew(Job);
+            }
+            else
+            {
+                testJobCompleted.Set();
+            }
+        }
+        Task.Factory.StartNew(Job);
+        Assert.True(testJobCompleted.WaitOne(ThreadTestHelpers.ExpectedMeasurableTimeoutMilliseconds));
+        e0.Set();
+    }
+
+    private static Action CreateRecursiveJob(int jobCount, int targetJobCount, AutoResetEvent testJobCompleted)
+    {
+        return () =>
+        {
+            if (jobCount == targetJobCount)
+            {
+                testJobCompleted.Set();
+            }
+            else
+            {
+                Task.Factory.StartNew(CreateRecursiveJob(jobCount + 1, targetJobCount, testJobCompleted));
+            }
+        };
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void RunJobsAfterThreadTimeout()
+    {
+        ManualResetEvent e0 = new ManualResetEvent(false);
+        AutoResetEvent jobsQueued = new AutoResetEvent(false);
+        AutoResetEvent testJobCompleted = new AutoResetEvent(false);
+        int count = 0;
+
+        for(int i = 0; i < Environment.ProcessorCount - 1; ++i)
+        {
+            Task.Factory.StartNew(() => {
+                if(Interlocked.Increment(ref count) == Environment.ProcessorCount - 1)
+                {
+                    jobsQueued.Set();
+                }
+                e0.WaitOne();
+            });
+        }
+        Assert.True(jobsQueued.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+        Task.Factory.StartNew(() => testJobCompleted.Set());
+        Assert.True(testJobCompleted.WaitOne(ThreadTestHelpers.ExpectedMeasurableTimeoutMilliseconds));
+        Console.Write("Sleeping to time out thread\n");
+        Thread.Sleep(21000);
+        Task.Factory.StartNew(() => testJobCompleted.Set());
+        Assert.True(testJobCompleted.WaitOne(ThreadTestHelpers.ExpectedMeasurableTimeoutMilliseconds));
+        e0.Set();
+        Console.Write("Sleeping to time out all threads\n");
+        Thread.Sleep(21000);
+        Task.Factory.StartNew(() => testJobCompleted.Set());
+        Assert.True(testJobCompleted.WaitOne(ThreadTestHelpers.ExpectedMeasurableTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void ThreadLocalWorkQueueDepletionTest()
+    {
+        ManualResetEvent e0 = new ManualResetEvent(false);
+        int count = 0;
+        int maxCount = Environment.ProcessorCount * 64;
+        object syncRoot = new object();
+        void Job()
+        {
+            if(Interlocked.Increment(ref count) >= maxCount)
+            {
+                e0.Set();
+            }
+            else
+            {
+                Task.Factory.StartNew(Job);
+                Task.Factory.StartNew(Job);
+            }
+        }
+        Task.Factory.StartNew(Job);
+        e0.WaitOne();
```

Comment:
Think this could be done a bit differently. Currently, it's only verifying that so many jobs are queued rather than completed.

Suggestion:
- Increment a counter before queuing each work item, ensure that exactly maxCount work items are queued in total
- Increment a counter before returning from a work item
- After queuing the task, also queue ProcessorCount * 64 work items using QueueUserWorkItem, they will go into the global queue, and for these work items increment the same counter as above before returning from the work item
- In any work item, when the return counter reaches ProcessorCount * 64 * 2, set an event, and verify that this event gets set at the end